CleverRaven/Cataclysm-DDA
# Refactor salvage_actor::cut_up

#### Summary
Bugfixes "Refactor salvage_actor::cut_up and fix minor bugs"


#### Purpose of change
This PR fixes #54657 which pointed out an issue with the formula for the returned weight.
I found another issue with the item loss, due to low fabrication etc., being applied multiple times.


#### Describe the solution
While fixing the issue with multiple materials, I decided to refactor the function to reduce code duplication and make things clearer.
The changes are mostly:

Validation:
* Remove try_to_cut_up and use only valid_to_cut_up so only one function is used for validation. Not perfectly happy with the solution but I still think two nearly identical functions will cause problems in the future.
* Make cut_up private and wrap it with try_to_cut_up which does the validation itself. An external call without validation is no longer possible
* Move remaining validation completely outside of cut_up instead of doing it twice.

Refactoring:
* Rename some vars for clarity (it -> cutter/cut, remaining_weight -> efficiency, salvage_to -> salvage)
* Extract `find_uncraft_recipe` to simplify `cut_up`. It attempts to find a crafting or disassembly recipe fitting some constraints.
* Extract `cut_up_component` into a recursive lambda. Makes the order of traversal clearer and passing arguments simpler.

Changes
* Lift the restriction on recipe weight. As the ingredients weigh more most of the time recipes were rarely used previously. Now efficiency is decreased while reversing the recipe.
* Prefer disassembly recipes to normal ones. Previously cutting items up could give more than disassembling
* Increase malus for low fab skill (1% -> 10%)
* Discard unsalvageable ~or count_by_charges~ item instead of distributing their weight to their materials


#### Describe alternatives you've considered
Not refactoring stuff and leaving things a whole lot less clear. It would be simpler to review I guess. Unfortunately changes to the indentation make the side-by-side diff basically useless.


#### Testing
The tests succeed and I've tested it with some specific items:
Item | Weight Original | Weight Salvage
-|-|-
Leather Vest | 1.00 | 0.96
Pair of Survivor Boots | 1.33 | 0.22
Scavenger Gear | 7.92 | 7.8
US Ballistic Vest | 2.91 | 2.62

Note that the boots can be disassembled to obtain a rubber sole. As  cut_up now prefers disassembly recipes when possible, a lot of weight is lost.



#### Additional context
I'm unsure about some potential problems and would like some feedback:
* The penalty for cutting up is 0% (besides rounding) for high fab + dex characters.
  * Decrease efficiency to a flat 75% + additional modifiers for fab/dex.
* Items with the NO_SALVAGE flag but with appropriate materials were still cut up.
  *  This is unintuitive. The whole item is discarded now.
* cut_up returns some weird values for the cost.
  * We could use time_to_cut_up for the cost instead. Seems more natural but I'm not sure what the value is exactly used for.

Ideally both fabrication and dexterity requirements would relate to the complexity of the item. Not sure how to do that. Do recipes exist for the majority of items? Then we could use their skill level to determine complexity. Could also use hardness of materials (and/or number of materials although that was just removed by this PR :) ). I think thats out-of-scope for this PR.
